### PR TITLE
REF-1068 Add additional lifecycle methods

### PR DESCRIPTION
**Pull request by georgelesica-wf**
[Jira Ticket](https://jira.atl.workiva.net/browse/REF-1068)

## Problem

Lux uses a pattern whereby they add their components in an `onDidLoadChildModule` method that gets called automagically when a child module is loaded. This seems to work out very nicely and results in clean, easy-to-read code.

However, since we don't support this directly in the SDK or in w_module it is left up to consumers to implement the boilerplate and many consumers might not pick up on it unless they read the Lux source code. Additionally, and this is fairly minor, the only way to wire this up at the level of an Experience is to listen to the appropriate stream, which is slightly complex due to the need to cancel stream subscriptions.

I originally set out to add this method to the `Experience` in the wdesk_sdk repo, but it was suggested that this method, and similar child lifecycle methods might be useful on the `LifecycleModule` itself.

## Solution

Add support for this pattern.

## Potential Regressions

None?

## Tests

Added and/or updated.

## QA / +10

1. Checkout the branch.
2. Update dependencies.
3. Unit tests should pass.
4. Pull this branch into the `wdesk_examples` repo and make sure everything still works.
5. Pull this branch into `docs`, get rid of their homebrew stream subscription that makes this pattern work, and verify that their app still works (specifically, you should be able to create a document). Here's a branch with these changes already made: https://github.com/Workiva/docs/tree/test-ref-1068.

## FYI / FYA / FYR

@trentgrover-wf @maxwellpeterson-wf @evanweible-wf @jayudey-wf @Workiva/ref-pp

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Raven
Number of Findings: 0

File: lib/src/lifecycle_module.dart [hunk 1]
```diff
@@ -80,12 +106,30 @@ abstract class LifecycleModule {
   /// Public method to async load a child module and register it
   /// for lifecycle management.
   Future loadChildModule(LifecycleModule newModule) {
+    if (_childModules.contains(newModule)) {
```
Comment (georgelesica-wf) [inline lib/src/lifecycle_module.dart]: This keeps us from listening to the child's event streams more than once and ensures that the subscriptions get fully canceled when the child module unloads.

**Comment by todbachman-wf** [comment]
+1

**Comment by bencampbell-wf** [comment]
+1

**Comment by bencampbell-wf** [comment]
@jayudey-wf ready for merge.

**Comment by evanweible-wf** [comment]
+1 Looks good

**Comment by jayudey-wf** [comment]
QA Resource Approval: +10

- [ ] Testing instruction
- [ ] Dev +1's
- [ ] Dev/QA +10 with detail of what was tested
 - performed and documented by Tod
- [ ] Unit test created/updated
- [ ] All unit tests pass 

Merging into master.

**Pull request merged**